source/distributor/service_impl.cc

```diff
@@ -0,0 +1,117 @@
+  std::unique_ptr<nighthawk::client::NighthawkService::Stub> stub;
+  std::shared_ptr<::grpc::Channel> channel;
+  channel = grpc::CreateChannel(fmt::format("{}:{}", service.socket_address().address(),
+                                            service.socket_address().port_value()),
+                                grpc::InsecureChannelCredentials());
+  stub = std::make_unique<nighthawk::client::NighthawkService::Stub>(channel);
+  return service_client_->PerformNighthawkBenchmark(stub.get(), request.start_request().options());
+}
+
+// Translates one or more backend response into a single reply message
+std::tuple<grpc::Status, nighthawk::DistributedResponse>
+NighthawkDistributorServiceImpl::handleRequest(
+    const ::nighthawk::DistributedRequest& request) const {
+  RELEASE_ASSERT(request.has_execution_request(), "request.has_execution_request()");
+  RELEASE_ASSERT(request.services_size() != 0, "services_size() == 0");
+  RELEASE_ASSERT(request.execution_request().has_start_request(), "no start_request");
+  RELEASE_ASSERT(request.execution_request().start_request().has_options(), "no options");
+  ENVOY_LOG(trace, "Handling execution request");
+
+  nighthawk::DistributedResponse response;
+  bool has_errors = false;
+  for (const envoy::config::core::v3::Address& service : request.services()) {
+    absl::StatusOr<nighthawk::client::ExecutionResponse> execution_response =
+        handleExecutionRequest(service, request.execution_request());
+    nighthawk::DistributedResponseFragment* response_fragment = response.add_fragment();
+    response_fragment->mutable_service()->MergeFrom(service);
+    if (execution_response.ok()) {
+      *response_fragment->mutable_execution_response() = execution_response.value();
+    } else {
+      response_fragment->mutable_error()->set_code(
+          static_cast<int>(execution_response.status().code()));
+      response_fragment->mutable_error()->set_message(
+          std::string("Distributed Execution Request failed: ") +
+          std::string(execution_response.status().message()));
+      has_errors = true;
+    }
+  }
+  return {has_errors
+              ? grpc::Status(grpc::StatusCode::INTERNAL, "One or more execution requests failed")
+              : grpc::Status::OK,
+          response};
+}
+
+::grpc::Status NighthawkDistributorServiceImpl::DistributedRequestStream(
+    ::grpc::ServerContext*,
+    ::grpc::ServerReaderWriter<::nighthawk::DistributedResponse, ::nighthawk::DistributedRequest>*
+        stream) {
+  nighthawk::DistributedRequest request;
+  ::grpc::Status status = grpc::Status::OK;
+  while (status.ok() && stream->Read(&request)) {
+    ENVOY_LOG(trace, "Inbound DistributedRequest {}", request.DebugString());
+    status = validateRequest(request);
+    if (status.ok()) {
+      std::tuple<grpc::Status, nighthawk::DistributedResponse> status_and_response =
+          handleRequest(request);
+      status = std::get<0>(status_and_response);
```

Comment:
if we know status = ok, this line has no purpose.

Reply:
Well this does break the while loop when status is not OK?